src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java

```diff
@@ -432,6 +434,10 @@ protected static String getNameIndex(String topic, int partition) {
     private boolean hasInitialized = false;
     private boolean forceCleanerFileReset = false;
 
+    // exactly once semantics
+    private final AtomicLong clientSequencer = new AtomicLong(-1);
+    private final AtomicLong serverSideOffset = new AtomicLong(-1);
```

Comment:
may be something like `offsetPersistedInSnowflake`

Can you also add a comment saying `this is source of truth` Also this value doesnt necessarily mean it is available to query but just a notification that snowflake has received this offset and has queued it?

Reply:
For verifying whether that offset is ingested, we still have to rely on insertReport api